Lib/hashlib.py

```diff
@@ -254,6 +255,122 @@ def prf(msg, inner=inner, outer=outer):
     pass
 
 
+import socket as _socket
+import binascii as _binascii
+
+class _LinuxKCAPI:
+    """Linux Kernel Crypto API (AF_ALG socket)
+    """
+
+    # TODO: retrieve info from AF_NETLINK, NETLINK_CRYPTO
+    _digests = {
+        'md5': (16, 64),
+        'sha1': (20, 64),
+        'sha224': (28, 64),
+        'sha256': (32, 64),
+        'sha384': (48, 128),
+        'sha512': (64, 128),
+    }
+    __slots__ = ("_digest", "_opsock")
+
+    def __init__(self, digest):
+        if digest not in self._digests:
+            raise ValueError(digest)
+        self._digest = digest
+
+    def _get_opsock(self, digest, key=None):
+        """Create KCAPI client socket
+
+        Algorithm and MAC key are configured on the server socket. accept()
+        creates a new client socket that consumes data. accept() on a client
+        socket creates an independent copy.
+        """
+        with _socket.socket(_socket.AF_ALG, _socket.SOCK_SEQPACKET, 0) as cfg:
+            binding = ("hash", digest if key is None else f"hmac({digest})")
+            try:
+                cfg.bind(binding)
+            except FileNotFoundError:
+                raise ValueError(binding)
+            if key is not None:
+                cfg.setsockopt(_socket.SOL_ALG, _socket.ALG_SET_KEY, key)
+            return cfg.accept()[0]
+
+    def __del__(self):
+        if getattr(self, "_opsock", None):
+            self._opsock.close()
+            self._opsock = None
+
+    def __repr__(self):
+        return f"<Linux KCAPI {self.name} at 0x{id(self):x}>"
+
+    @property
+    def digest_size(self):
+        return self._digests[self._digest][0]
+
+    @property
+    def block_size(self):
+        return self._digests[self._digest][1]
+
+    def update(self, data):
+        self._opsock.sendall(data, _socket.MSG_MORE)
+
+    def copy(self):
```

Comment:
Add a `__getstate__` method that raises to prevent `pickle.dumps` or `copy.copy` from working on these.